cvatjs/src/annotations.js

```diff
@@ -287,73 +336,402 @@
-                        leftPosition.points[2] + positionOffset[2] * offset,
-                        leftPosition.points[3] + positionOffset[3] * offset,
-                    ],
-                    occluded: leftPosition.occluded,
-                    outside: leftPosition.outside,
-                    zOrder: leftPosition.zOrder,
-                };
+            function normalize(points, box) {
+                const normalized = [];
+                const width = box.xmax - box.xmin;
+                const height = box.ymax - box.ymin;
+
+                for (let i = 0; i < points.length; i += 2) {
+                    normalized.push(
+                        (points[i] - box.xmin) / width,
+                        (points[i + 1] - box.ymin) / height,
+                    );
+                }
+
+                return normalized;
             }
 
-            if (rightPosition) {
-                return {
-                    points: [...rightPosition.points],
-                    occluded: rightPosition.occluded,
-                    outside: true,
-                    zOrder: 0,
-                };
+            function denormalize(points, box) {
+                const denormalized = [];
+                const width = box.xmax - box.xmin;
+                const height = box.ymax - box.ymin;
+
+                for (let i = 0; i < points.length; i += 2) {
+                    denormalized.push(
+                        points[i] * width + box.xmin,
+                        points[i + 1] * height + box.ymin,
+                    );
+                }
+
+                return denormalized;
             }
 
-            if (leftPosition) {
-                return {
-                    points: [...leftPosition.points],
-                    occluded: leftPosition.occluded,
-                    outside: leftPosition.outside,
-                    zOrder: 0,
+            function toPoints(array) {
+                const points = [];
+                for (let i = 0; i < array.length; i += 2) {
+                    points.push({
+                        x: array[i],
+                        y: array[i + 1],
+                    });
+                }
+
+                return points;
+            }
+
+            function toArray(points) {
+                const array = [];
+                for (const point of points) {
+                    array.push(point.x, point.y);
+                }
+
+                return array;
+            }
+
+            function computeDistances(source, target) {
+                const distances = {};
+                for (let i = 0; i < source.length; i++) {
+                    distances[i] = distances[i] || {};
+                    for (let j = 0; j < target.length; j++) {
+                        const dx = source[i].x - target[j].x;
+                        const dy = source[i].y - target[j].y;
+
+                        distances[i][j] = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2));
```

Comment:
sqrt isn't necessary here (you will compare these values, you don't need absolute value). Also if the method is slow in the future we can use L1 distance = abs(dx) + abs(dy)